src/code/z_camera.c

```diff
@@ -7602,19 +7607,22 @@ Vec3s Camera_Update(Camera* camera) {
         return camera->inputDir;
     }
 
-    // setting bgId to the ret of Quake_Calc, and checking that
-    // is required, it doesn't make too much sense though.
-    if ((bgId = Quake_Calc(camera, &quake), bgId != 0) && (camera->setting != CAM_SET_TURN_AROUND)) {
-        viewAt.x = camera->at.x + quake.atOffset.x;
-        viewAt.y = camera->at.y + quake.atOffset.y;
-        viewAt.z = camera->at.z + quake.atOffset.z;
-        viewEye.x = camera->eye.x + quake.eyeOffset.x;
-        viewEye.y = camera->eye.y + quake.eyeOffset.y;
-        viewEye.z = camera->eye.z + quake.eyeOffset.z;
+    // setting bgId to the ret of Quake_Update is required.
+    bgId = numQuakesApplied = Quake_Update(camera, &quakeCamData);
+
+    if ((numQuakesApplied != 0) && (camera->setting != CAM_SET_TURN_AROUND)) {
+        viewAt.x = camera->at.x + quakeCamData.atOffset.x;
+        viewAt.y = camera->at.y + quakeCamData.atOffset.y;
+        viewAt.z = camera->at.z + quakeCamData.atOffset.z;
+
+        viewEye.x = camera->eye.x + quakeCamData.eyeOffset.x;
+        viewEye.y = camera->eye.y + quakeCamData.eyeOffset.y;
+        viewEye.z = camera->eye.z + quakeCamData.eyeOffset.z;
+
         OLib_Vec3fDiffToVecSphGeo(&eyeAtAngle, &viewEye, &viewAt);
-        Camera_CalcUpFromPitchYawRoll(&viewUp, eyeAtAngle.pitch + quake.rotZ, eyeAtAngle.yaw + quake.unk_1A,
-                                      camera->roll);
-        viewFov = camera->fov + CAM_BINANG_TO_DEG(quake.zoom);
+        Camera_CalcUpFromPitchYawRoll(&viewUp, eyeAtAngle.pitch + quakeCamData.rollFromPitch,
+                                      eyeAtAngle.yaw + quakeCamData.rollFromYaw, camera->roll);
```

Comment:
so that's because these offsets are only used to compute the view up, the changed pitch/yaw only affects viewUp
the forwards direction is still given by eye and at, which don't depend on those pitch/yaw offsets
hence why the end results only looks like roll is being applied

now, this is very sketchy because up and forwards are supposed to be orthogonal (perpendicular) but this doesn't respect that
with your debugger, try setting pitch to 0x4000, it will make forwards and up almost aligned and you will find the view behaves... oddly (seizure alert!)

in cases were up and forwards aren't too aligned though, it's fine thanks to how `guLookAtF` works (it computes viewRight from forwards and up, then computes a proper up from right and forwards)

maybe good names would be upPitchOffset and upYawOffset

Reply:
I forgot to post this addition to your debugger btw
```c
    {
        Vec3f viewForwards;
        f32 norm;

        norm = Math_Vec3f_DistXYZAndStoreDiff(&viewAt, &viewEye, &viewForwards);
        Math_Vec3f_Scale(&viewForwards, 1 / norm);

        osSyncPrintf("pitch 0x%04X yaw 0x%04X\n", (u16)sCamDebugPitchOffset, (u16)sCamDebugYawOffset);
        osSyncPrintf("viewUp       %.3f %.3f %.3f\n", viewUp.x, viewUp.y, viewUp.z);
        osSyncPrintf("viewForwards %.3f %.3f %.3f\n", viewForwards.x, viewForwards.y, viewForwards.z);
        osSyncPrintf("dot = %f\n", DOTXYZ(viewUp, viewForwards));
    }
```

Reply:
also
> thanks to how guLookAtF works (it computes viewRight from forwards and up, then computes a proper up from right and forwards)

this explains why we see "roll", the effective ends up always being orthogonal to forwards

Reply:
Curiously, they change it in MM as well to only have 1 variable:
`Camera_CalcUpFromPitchYawRoll(&viewUp, eyeAtAngle.pitch, eyeAtAngle.yaw, camera->roll + quakeCamData.rollOffset);`

I'd be fine with `upPitchOffset` and `upYawOffset` to associate it with the up vector. Since these variables trace back to the `roll` in:
`u32 Quake_SetPerturbations(s16 index, s16 yOffset, s16 xOffset, s16 fov, s16 roll);`
should `roll` also reflect this somehow? It might be confusing from a user perspective though

Reply:
`roll` only ends up used for upPitchOffset

it's just a very weird implementation all around, as changing only the pitch (and not yaw/roll) of up serves no purpose, since the part of up that's collinear with forwards ends up discarded in guLookAtF, and pitch on up only makes it move towards forwards on the same plane (this can be observed with your debugger, keep yaw at 0 and set pitch to anything (except values that make the lookAt jank like 0x4000))
what the pitch offset actually does though, is that it changes how the yaw affects up. for example if pitch is 0, up is +y and the yaw does nothing. If pitch is larger, up moves away from +y and yaw (which rotates around that +y) has an actual effect.
for example with yaw=0x4000, up ends up looking like normal up but with a roll of "pitch", since the angle +y was pitched by ends up left of the plane formed by +y and forwards

My best guess on why they did this is they just modified random things on the view until it looked not too jank and then moved on
Which imo motivates naming things after what they change, at some level

Naming the argument currently named "roll" is harder though, technically I guess it'd be `upPitchOffsetMagnitude` but that's indeed completely useless to the user

side note
```c
// Quake_UpdateShakeInfo
shake->rollFromYaw = 0x8000 * y;
shake->rollFromPitch = req->roll * y;
```
so `rollFromPitch`/upPitchOffset is more or less "roll" when `rollFromYaw`/upYawOffset is more or less 0x4000, i.e. y=0.5, i.e. `rollFromPitch = req->roll * 0.5` YEP

NotLikeThis I don't know
probably keep roll somewhere with a comment when it turns into pitch/yaw offsets about... stuff

Reply:
```c
if (camData->rollFromPitch < shake.rollFromPitch) {
    camData->rollFromPitch = shake.rollFromPitch;
    camData->rollFromYaw = shake.rollFromYaw;
}
```
Should be noted that they are still connected somehow? In that `rollFromYaw` is only passed on from shake if the new `rollFromPitch` (from updating shake) is larger than the current `rollFromPitch` value (from the previous camData)

Reply:
`rollFromYaw` isn't really configurable though, it depends almost exclusively on the quake type, it's just `0x8000 * y`
I think it's not user-facing enough to be described in docs that aren't comments inside the functions

taking its max also wouldn't work, it'd just approach 0x8000 and doing nothing (it's most effective at 0x4000)

Reply:
The "end" effect that the user sees "appears" to be a roll, although weirdly implemented, since it's not actually changing the roll in any way. My current inclination is to still use `roll` in `Quake_SetPerturbations`, but then use the technical `upPitchOffsetMagnitude` for all internal variables past that stage? With a comment explaining why:
`req->upPitchOffsetMagnitude = roll; // explain here`
inside `Quake_SetPerturbations`, but idk. Need to think about it a bit more

Reply:
yeah that's what I'm thinking too